specs/WinRT/MakeMSIX-spec.md

```diff
@@ -0,0 +1,959 @@
+            RRF_RT_REG_SZ,
+            nullptr /* pdwType */,
+            &productVersionBuffer,
+            &productVersionBufferLength));
+        std::wstring productVersionString{ productVersionBuffer };
+        // The install path of the platform sdk in the filesystem always uses version format x.x.x.x
+        // even though the version string in the registry may be stored as x.x if the full version is
+        // x.x.0.0 
+        auto versionPartsFound = std::count(productVersionString.begin(), productVersionString.end(), '.') + 1;
+        INT64 versionPartsRequired = 4;
+        INT64 versionDigitsToAdd = versionPartsRequired - versionPartsFound;
+        for (auto i = 0; i < versionDigitsToAdd; i++)
+        {
+            productVersionString += L".0";
+        }
+        std::wstring processorArchitecture{};
+        SYSTEM_INFO systemInfo;
+        GetNativeSystemInfo(&systemInfo);
+        switch (systemInfo.wProcessorArchitecture)
+        {
+        case PROCESSOR_ARCHITECTURE_AMD64:
+            processorArchitecture += L"x64";
+            break;
+        case PROCESSOR_ARCHITECTURE_ARM:
+            processorArchitecture += L"arm";
+            break;
+        case PROCESSOR_ARCHITECTURE_ARM64:
+            processorArchitecture += L"arm64";
+            break;
+        case PROCESSOR_ARCHITECTURE_INTEL:
+            processorArchitecture += L"x86";
+            break;
+        case PROCESSOR_ARCHITECTURE_IA64:
+        case PROCESSOR_ARCHITECTURE_UNKNOWN:
+            winrt::throw_hresult(E_UNEXPECTED);
+        }
+
+        std::filesystem::path platformSDKExecutablePath = installPathString;
+        platformSDKExecutablePath /= L"bin";
+        platformSDKExecutablePath /= productVersionString;
+        platformSDKExecutablePath /= processorArchitecture;
+
+        return platformSDKExecutablePath;
+    }
+
+    void LaunchSignToolWithArguments(std::wstring arguments)
+    {
+        std::filesystem::path signToolPath = GetPlatformSDKPath();
+        signToolPath /= L"signtool.exe";
+
+        std::wstring commandLine{ signToolPath };
+        commandLine += L" " + arguments;
+        DWORD exitCode{};
+        winrt::check_hresult(CreateProcessAndWaitForExitCode(commandLine, exitCode));
+        if (exitCode == EXIT_FAILURE)
+        {
+            winrt::throw_hresult(E_FAIL);
+        }
+    }
+
+    void Example_SignPackage(std::wstring pfxFile, std::wstring packageFile)
+    {
+        std::wstring signToolArguments;
+        signToolArguments.append(L"sign /fd SHA256 /f ");
+        signToolArguments.append(pfxFile);
+        signToolArguments.append(L" " + packageFile);
+        LaunchSignToolWithArguments(signToolArguments);
+    }
+```
+
+# 5. API Details
+
+```c#
+// Copyright (c) Microsoft Corporation and Contributors.
+// Licensed under the MIT License.
+
+namespace Microsoft.Kozani.MakeMSIX
+{
+    [contractversion(1)]
+    apicontract MakeMSIXContract{};
+
+    /// Options for creating a package
```

Comment:
Ohhhhhh. PackOptions are options for the Pack() command to create a package vs CreateKozaniPackageOptions are options to the ??? command to create a LocalVariant package. This was confusing to me (think about that statement..) but OK.

Q: What options differ? Is creating a LocalVariant package the same as creating a package plus additional options?

What options differ?

If there's `Pack(input, packOptions)` would it make more sense to have `CreateKozani(input, packOptions, createKozaniOptions)`? If CreateKozani == Pack + additional behavior and options...

Reply:
Multiple comments dealing with roughly similar questions - choosing this one for response. The top level question I think is whether to merge kozani package creation into the other apis as an option, or to have it as a separate thing. As you noticed, the other apis all take an input of a specific format (i.e. CreatePackage requires a layout directory), an output that they always write in a specific format, an options object, and then perform one operation on the input file to produce the output.

Creation of Kozani packages could follow that model. A caller would call ExtractBundle, iterate through msix files and call ExtractPackage,  iterate directories and call CreatePackage with something like MakeKozaniPackage(true) and other options in CreatePackageOptions like RemoveExtensions(true), call CreateBundle with any Kozani specific options.

For Kozani though I think we know that by far the most common operation that people want to do is to convert from an existing package to Kozani. They may have a layout directory, or a package, or a bundle - and they just want that same thing in Kozani form. In some cases they may have large groups of packages, some of which are bundles and others of which are packages, that they want to convert. In this way, creation of a kozani package is more like an operation like AddPackage, where any type of package format (layout, package, bundle) can be handled internally, than it is like the other apis in this spec (like CreatePackage) where the input and output are always one format. 
Forcing people to unpack the entire package just to recreate it is a waste of processor time and disk space, and forcing them to special case bundles vs packages adds unnecessary code complexity for this one most common scenario. Of course those operations are all available if there's some custom manifest changes the caller wants to write.

In other words, since all the other apis transform an existing package or set of packages to some other format, whereas Kozani package creation is mostly intended to transform an existing package into a different version of the same package format, the only way to make Kozani package creation fit into the existing api space is to require callers to first transform the package out of the format they want, so that they can call another api with a kozani option to transform it back into the format they want.

Since I'm proposing having CreateKozaniPackage handle all package types internally, the equivalent option set is more like CreateKozaniOptions == CreatePackageOptions + CreateBundleOptions + additional kozani specific options (+ (potentially) CreateMountableImageOptions). Since these are still just packages I would mostly assume anything in the "additional kozani specific options" could be moved into either\both of CreatePackageOptions or CreateBundleOptions, just that some options are very unlikely to be useful for those operations and so seem likely to do nothing more than add confusion for people. As an example, in the update to the spec I just pushed I added "boolean RemoveExtensions" as field on the CreateKozaniPackageOptions. There is no inherent reason this couldn't be in CreatePackageOptions as well.

But I think the intention of having CreateKozaniPackage handle all types of packages internally does mean it makes sense to have its own options object. There may be cause to add the OutputFormat option you suggested in another comment, which wouldn't make sense for the other apis. Or there may be differences in behavior between operations using the same options. For example, since main packages could all have different versions and a bundle manifest isn't required for bundling, CreateBundle defaults to creating a bundle version based on timestamp if Version is not explicitly specified. But CreateKozaniPackage does require an existing package, so it makes more sense to maintain the version from the existing manifest if Version is not set.

Reply:
> the intention of having CreateKozaniPackage handle all types of packages internally does mean it makes sense to have its own options object

+1